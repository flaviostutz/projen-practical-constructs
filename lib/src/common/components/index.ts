export * from './base-tooling';
export * from './common-target-type';
export * from './common-targets';
export * from './release-tasks';
export * from './publish';
