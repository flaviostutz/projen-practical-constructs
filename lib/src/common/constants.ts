export const PROJEN_VERSION = '0.91.13';
export const NODE_VERSION = '20.16.0';
export const MONOTAG_VERSION = '1.23.0';
