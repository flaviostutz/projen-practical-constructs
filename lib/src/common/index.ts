export * from './components';
export * from './files';
export * from './types';
