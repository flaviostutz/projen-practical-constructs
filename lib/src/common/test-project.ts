import { mkdtempSync } from 'node:fs';
import { tmpdir } from 'node:os';
// eslint-disable-next-line unicorn/import-style
import { join } from 'node:path';

import { Project } from 'projen';

export class TestProject extends Project {
  constructor() {
    const outdir = mkdtempSync(join(tmpdir(), 'test-'));
    super({ name: 'test', outdir });
  }
}
