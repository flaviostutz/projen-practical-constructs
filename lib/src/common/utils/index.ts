export * from './ts2jsii';
export * from './monotag';
