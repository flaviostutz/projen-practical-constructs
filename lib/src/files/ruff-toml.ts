/* eslint-disable no-restricted-syntax */
import { Project, TomlFile } from 'projen';

/**
 * ruff.toml synthetisation
 */
export class RuffTomlFile extends TomlFile {
  constructor(project: Project, opts?: RuffTomlFileOptions) {
    const optsWithDefaults = getOptionsWithDefaults(opts);
    super(project, 'ruff.toml', {
      obj: {
        'cache-dir': '.cache/.ruff_cache',
        'show-fixes': true,
        src: ['src', 'tests'],
        'target-version': optsWithDefaults.targetPythonVersion,
        'unsafe-fixes': optsWithDefaults.unsafeFixes,
        format: {
          'docstring-code-format': true,
        },
        lint: {
          ignore: optsWithDefaults.ignoreRules,
          select: optsWithDefaults.selectRules,
          mccabe: {
            'max-complexity': optsWithDefaults.mccabeMaxComplexity,
          },
          'per-file-ignores': optsWithDefaults.perFileIgnores,
        },
      },
    });
  }
}

const getOptionsWithDefaults = (opts?: RuffTomlFileOptions): Required<RuffTomlFileOptions> => {
  const defaultOptions = {
    unsafeFixes: false,
    targetPythonVersion: 'py313',
    ignoreRules: [
      'TD003',
      'TRY002',
      'TRY003',
      'D100',
      'D203',
      'D213',
      'D400',
      'D415',
      'COM812',
      'ISC001',
      'W191',
      'E111',
      'E114',
      'E117',
      'D206',
      'D300',
      'Q000',
      'Q001',
      'Q002',
      'Q003',
      'COM812',
      'COM819',
      'ISC001',
      'ISC002',
    ],
    selectRules: [
      'E',
      'F',
      'UP',
      'B',
      'SIM',
      'I',
      'W',
      'C90',
      'N',
      'D',
      'ASYNC',
      'ANN',
      'S',
      'FBT003',
      'A',
      'COM',
      'C4',
      'DTZ',
      'T10',
      'DJ',
      'EXE',
      'FA',
      'ISC',
      'ICN',
      'LOG',
      'G',
      'INP',
      'PIE',
      'T20',
      'PYI',
      'PT',
      'Q',
      'RSE',
      'RET',
      'SLF',
      'TC',
      'INT',
      'ARG',
      'PTH',
      'TD',
      'ERA',
      'PD',
      'PGH',
      'PL',
      'TRY',
      'FLY',
      'NPY',
      'FAST',
      'PERF',
      'FURB',
      'RUF',
    ],
    addToExistingRules: true,
    perFileIgnores: {
      'tests/*': ['D', 'S101'],
    },
    mccabeMaxComplexity: 14,
    ...opts,
  };

  // add provided rules to default rules
  if (defaultOptions.addToExistingRules) {
    defaultOptions.ignoreRules = [...defaultOptions.ignoreRules, ...(opts?.ignoreRules ?? [])];
    defaultOptions.selectRules = [...defaultOptions.selectRules, ...(opts?.selectRules ?? [])];
    defaultOptions.perFileIgnores = {
      ...defaultOptions.perFileIgnores,
      ...(opts?.perFileIgnores ?? {}),
    };
    // replace rules with the provided ones
  } else {
    defaultOptions.ignoreRules = opts?.ignoreRules ?? defaultOptions.ignoreRules;
    defaultOptions.selectRules = opts?.selectRules ?? defaultOptions.selectRules;
    defaultOptions.perFileIgnores = opts?.perFileIgnores ?? defaultOptions.perFileIgnores;
  }

  return defaultOptions;
};

export interface RuffTomlFileOptions {
  /**
   * @default py313
   */
  readonly targetPythonVersion?: string;
  /**
   * @default pre-selected set of rules
   */
  readonly ignoreRules?: string[];
  /**
   * Add rules defined here to the ignoreRules and selectRules (on top of the default rules) or replace them by this contents?
   * @default true
   */
  readonly addToExistingRules?: boolean;
  /**
   * @default pre-selected set of rules
   */
  readonly selectRules?: string[];
  /**
   * @default ignore doc-string code format for test files
   */
  readonly perFileIgnores?: Record<string, string[]>;
  /**
   * @default false
   */
  readonly unsafeFixes?: boolean;
  /**
   * @default 14
   */
  readonly mccabeMaxComplexity?: number;
}
