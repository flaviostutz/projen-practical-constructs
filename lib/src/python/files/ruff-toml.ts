/* eslint-disable functional/immutable-data */
/* eslint-disable no-restricted-syntax */
import { Project, TomlFile } from 'projen';

/**
 * ruff.toml synthetisation
 */
export class RuffTomlFile extends TomlFile {
  constructor(project: Project, opts?: RuffTomlFileOptions) {
    const optsWithDefaults = getOptionsWithDefaults(opts);
    super(project, 'ruff.toml', {
      obj: {
        'cache-dir': '.cache/.ruff_cache',
        'show-fixes': true,
        src: ['src', 'tests'],
        'target-version': optsWithDefaults.targetPythonVersion,
        'unsafe-fixes': optsWithDefaults.unsafeFixes,
        format: {
          'docstring-code-format': true,
        },
        lint: {
          ignore: optsWithDefaults.ignoreRules,
          select: optsWithDefaults.selectRules,
          mccabe: {
            'max-complexity': optsWithDefaults.mccabeMaxComplexity,
          },
          'per-file-ignores': optsWithDefaults.perFileIgnores,
        },
      },
    });
  }
}

const getOptionsWithDefaults = (opts?: RuffTomlFileOptions): Required<RuffTomlFileOptions> => {
  const defaultRules = {
    ignoreRules: [
      'TD003',
      'TRY002',
      'TRY003',
      'D100',
      'D203',
      'D213',
      'D400',
      'D415',
      'COM812',
      'ISC001',
      'W191',
      'E111',
      'E114',
      'E117',
      'D206',
      'D300',
      'Q000',
      'Q001',
      'Q002',
      'Q003',
      'COM812',
      'COM819',
      'ISC001',
      'ISC002',
    ],
    selectRules: [
      'E',
      'F',
      'UP',
      'B',
      'SIM',
      'I',
      'W',
      'C90',
      'N',
      'D',
      'ASYNC',
      'ANN',
      'S',
      'FBT003',
      'A',
      'COM',
      'C4',
      'DTZ',
      'T10',
      'DJ',
      'EXE',
      'FA',
      'ISC',
      'ICN',
      'LOG',
      'G',
      'INP',
      'PIE',
      'T20',
      'PYI',
      'PT',
      'Q',
      'RSE',
      'RET',
      'SLF',
      'TC',
      'INT',
      'ARG',
      'PTH',
      'TD',
      'ERA',
      'PD',
      'PGH',
      'PL',
      'TRY',
      'FLY',
      'NPY',
      'FAST',
      'PERF',
      'FURB',
      'RUF',
    ],
    perFileIgnores: {
      'tests/*': ['D', 'S101', 'INP001'],
    },
  };

  const defaultOptions = {
    unsafeFixes: false,
    targetPythonVersion: 'py313',
    addToExistingRules: true,
    mccabeMaxComplexity: 14,
    ignoreRules: [],
    selectRules: [],
    perFileIgnores: {},
    ...opts,
  };

  if (opts?.addToExistingRules) {
    defaultOptions.ignoreRules = [
      ...defaultRules.ignoreRules,
      ...(defaultOptions.ignoreRules ?? []),
    ];
    defaultOptions.selectRules = [
      ...defaultRules.selectRules,
      ...(defaultOptions.selectRules ?? []),
    ];
    defaultOptions.perFileIgnores = {
      ...defaultRules.perFileIgnores,
      ...defaultOptions.perFileIgnores,
    };
  }
  return defaultOptions;
};

export interface RuffTomlFileOptions {
  /**
   * @default py313
   */
  readonly targetPythonVersion?: string;
  /**
   * @default pre-selected set of rules
   */
  readonly ignoreRules?: string[];
  /**
   * Add rules defined here to the ignoreRules and selectRules (on top of the default rules) or replace them by this contents?
   * @default true
   */
  readonly addToExistingRules?: boolean;
  /**
   * @default pre-selected set of rules
   */
  readonly selectRules?: string[];
  /**
   * @default ignore doc-string code format for test files
   */
  readonly perFileIgnores?: Record<string, string[]>;
  /**
   * @default false
   */
  readonly unsafeFixes?: boolean;
  /**
   * @default 14
   */
  readonly mccabeMaxComplexity?: number;
}
