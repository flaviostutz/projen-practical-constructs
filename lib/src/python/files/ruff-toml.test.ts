/* eslint-disable no-new */
import { Testing } from 'projen';

import { TestProject } from '../../common/test-project';

import { RuffTomlFile } from './ruff-toml';

describe('RuffTomlFile', () => {
  it('synthesizes ruff.toml with default options', () => {
    const prj = new TestProject();
    new RuffTomlFile(prj, {
      /* ...test options... */
    });
    const out = Testing.synth(prj);
    expect(out['ruff.toml']).toMatchSnapshot();
  });

  it('synthesizes ruff.toml with custom options', () => {
    const prj = new TestProject();
    new RuffTomlFile(prj, {
      targetPythonVersion: 'py310',
      unsafeFixes: true,
      mccabeMaxComplexity: 10,
      ignoreRules: ['E501'],
      selectRules: ['F401'],
      perFileIgnores: {
        'src/*': ['E302'],
      },
    });
    const out = Testing.synth(prj);
    expect(out['ruff.toml']).toMatchSnapshot();
  });
});
