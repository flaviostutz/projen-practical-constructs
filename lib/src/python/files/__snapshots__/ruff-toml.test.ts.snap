// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RuffTomlFile synthesizes ruff.toml with custom options 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

cache-dir = ".cache/.ruff_cache"
show-fixes = true
src = [ "src", "tests" ]
target-version = "py310"
unsafe-fixes = true

[format]
docstring-code-format = true

[lint]
ignore = [ "E501" ]
select = [ "F401" ]

  [lint.mccabe]
  max-complexity = 10

  [lint.per-file-ignores]
  "src/*" = [ "E302" ]
"
`;

exports[`RuffTomlFile synthesizes ruff.toml with default options 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".

cache-dir = ".cache/.ruff_cache"
show-fixes = true
src = [ "src", "tests" ]
target-version = "py313"
unsafe-fixes = false

[format]
docstring-code-format = true

[lint]
ignore = [ ]
select = [ ]
per-file-ignores = { }

  [lint.mccabe]
  max-complexity = 14
"
`;
